storage/tick.go

```diff
@@ -0,0 +1,161 @@
+	errTickInProgress  = errors.New("another tick is in progress")
+	errTickCancelled   = errors.New("tick is cancelled")
+)
+
+type tickManagerMetrics struct {
+	tickDuration       tally.Timer
+	tickCancelled      tally.Counter
+	tickDeadlineMissed tally.Counter
+	tickDeadlineMet    tally.Counter
+}
+
+func newTickManagerMetrics(scope tally.Scope) tickManagerMetrics {
+	return tickManagerMetrics{
+		tickDuration:       scope.Timer("duration"),
+		tickCancelled:      scope.Counter("cancelled"),
+		tickDeadlineMissed: scope.Counter("deadline.missed"),
+		tickDeadlineMet:    scope.Counter("deadline.met"),
+	}
+}
+
+type tickManager struct {
+	sync.RWMutex
+
+	database database
+	opts     Options
+	nowFn    clock.NowFn
+	sleepFn  sleepFn
+
+	metrics tickManagerMetrics
+	c       context.Cancellable
+	tokenCh chan struct{}
+}
+
+func newTickManager(database database) databaseTickManager {
+	opts := database.Options()
+	scope := opts.InstrumentOptions().MetricsScope().SubScope("tick")
+	tokenCh := make(chan struct{}, 1)
+	tokenCh <- struct{}{}
+
+	return &tickManager{
+		database: database,
+		opts:     opts,
+		nowFn:    opts.ClockOptions().NowFn(),
+		sleepFn:  time.Sleep,
+		metrics:  newTickManagerMetrics(scope),
+		c:        context.NewCancellable(),
+		tokenCh:  tokenCh,
+	}
+}
+
+func (mgr *tickManager) Tick(softDeadline time.Duration, forceType forceType) error {
+	if forceType == force {
+		acquired := false
+		waiter := time.NewTicker(tokenCheckInterval)
+		// NB(xichen): cancellation is done in a loop so if there are multiple
+		// forced ticks, their cancellations don't get reset when token is acquired.
+		for !acquired {
+			select {
+			case <-mgr.tokenCh:
+				acquired = true
+			case <-waiter.C:
+				mgr.c.Cancel()
+			}
+		}
+		waiter.Stop()
+	} else {
+		select {
+		case <-mgr.tokenCh:
+		default:
+			return errTickInProgress
+		}
+	}
+
+	// Release the token
+	defer func() { mgr.tokenCh <- struct{}{} }()
+
+	// Now we acquired the token, reset the cancellable
+	mgr.c.Reset()
+	namespaces := mgr.database.getOwnedNamespaces()
+	if len(namespaces) == 0 {
+		return errEmptyNamespaces
+	}
+
+	// Begin ticking
+	start := mgr.nowFn()
+	sizes := make([]int64, 0, len(namespaces))
+	totalSize := int64(0)
+	for _, n := range namespaces {
+		size := n.NumSeries()
+		sizes = append(sizes, size)
+		totalSize += size
+	}
+	for i, n := range namespaces {
+		deadline := float64(softDeadline) * (float64(sizes[i]) / float64(totalSize))
+		n.Tick(mgr.c, time.Duration(deadline))
+	}
+
+	end := mgr.nowFn()
+	duration := end.Sub(start)
+	mgr.metrics.tickDuration.Record(duration)
+
+	if duration > softDeadline {
+		mgr.metrics.tickDeadlineMissed.Inc(1)
+	} else {
+		mgr.metrics.tickDeadlineMet.Inc(1)
+
+		// Throttle to reduce locking overhead during ticking
+		for d := time.Duration(0); d < softDeadline-duration; d += cancellationCheckInterval {
+			if mgr.c.IsCancelled() {
+				break
+			}
+			mgr.sleepFn(cancellationCheckInterval)
+		}
```

Comment:
Maybe just remove this extra sleep cycle.  It's going to be such a micro sleep as we would only sleep here if it is the last batch of series being slight faster than it should be (since we check and sleep every batch of series if we are ahead).  i.e. if we have 6MM series and we do batches of 128 and we do a 10min tick, then the most we would be ahead of the tick here would be `= 10mins / (6MM/128) = 12.8 milliseconds`.